### ISSUE TITLE
OperationTimeoutException when calling get on task future after hazelcast.operation.call.timeout.millis

### ISSUE THREAD

**Issue by atschofen** [hazelcast/hazelcast#4398]
I am executing some tests with a build of Hazelcast 3.4.1-dev from December 22 2014. I discovered a problem with long-running tasks similar to issue #1949. It seems to be new in 3.4/3.4.1.

The problem is the following: If you execute a task that takes at least two times as long as the configured hazelcast.operation.call.timeout.millis, and you call get on a remote future sometime after two times the timeout, an OperationTimeoutException (message "No response for x ms. Aborting invocation!...") is thrown. This can easily happen if you execute a long-running task and call get on all futures, since the first get will block until the task has finished on the corresponding node.

The following unit test can be used to reproduce the problem. It fails with the beforementioned Hazelcast 3.4.1 build, but not with e.g., 3.3.4:

```java
import com.hazelcast.config.Config;
import com.hazelcast.config.XmlConfigBuilder;
import com.hazelcast.core.Hazelcast;
import com.hazelcast.core.HazelcastInstance;
import com.hazelcast.core.IExecutorService;
import com.hazelcast.core.Member;
import com.nm.test.hazelcast.TestHazelcast;
import org.apache.log4j.BasicConfigurator;
import org.apache.log4j.Logger;
import java.io.Serializable;
import java.util.Map;
import java.util.concurrent.Callable;
import java.util.concurrent.Future;
import junit.framework.TestCase;

public class TestTask2 extends TestCase {

	private static final Logger logger = Logger.getLogger(TestTask1.class);

	@Override
	protected void setUp() throws Exception {

		// configure logging
		if (!TestHazelcast.loggingInitialized) {
			TestHazelcast.loggingInitialized = true;
			BasicConfigurator.configure();
		}
	}

	public void testTask2() throws Exception {

		// create hazelcast config
		Config config = new XmlConfigBuilder().build();
		config.setProperty("hazelcast.logging.type", "log4j");

		// set call timeout to 1 second to make the problem appear quicker
		config.setProperty("hazelcast.operation.call.timeout.millis", "2000");

		// create two Hazelcast instances
		HazelcastInstance hcInstance1 = Hazelcast.newHazelcastInstance(config);
		HazelcastInstance hcInstance2 = Hazelcast.newHazelcastInstance(config);
		try {

			// submit the long-running task
			IExecutorService executorService = hcInstance1.getExecutorService("default");
			Map<Member, Future<Object>> futures = executorService.submitToAllMembers(new SleepingTask());

			// call get on the "remote" member
			Member remoteMember = hcInstance2.getCluster().getLocalMember();
			// wait for 5 seconds
			// when we call get on the future after sleeping, the OperationTimeoutException will be raised in Hazelcast 3.4.1-20141222 dev
			Thread.sleep(5000);
			String result = (String) futures.get(remoteMember).get();

			assertEquals("Success", result);

		} finally {
			// shutdown Hazelcast instances
			hcInstance1.getLifecycleService().shutdown();
			hcInstance2.getLifecycleService().shutdown();
		}

		logger.info("Test done.");
	}

	private static class SleepingTask implements Callable<Object>, Serializable {

		private static final long serialVersionUID = 1L;

		@Override
		public Object call() throws Exception {
			Thread.sleep(6000);
			return "Success";
		}

	}

}
```

Thanks for taking a look and cheers,
Andreas

**Comment by mdogan** [comment]
This happens only if more than `2 x timeout` passes before calling `future.get()`. 
```java
Thread.sleep(5000);
String result = (String) futures.get(remoteMember).get();
```

Thanks for the report.

**Comment by lukasblu** [comment]
Hi,

I am a bit worried about this issue, because in our application it is very likely that someone will write a distributed task that takes longer than 2 minutes. As we check for the results in a loop calling get() on each Future object, we would then run into this problem.

Therefore, I have two questions:

1. Did you already have a chance to look at this? Is it something that can be fixed in the next few days?

2. If this is not so easy to fix, what consequences - in general - would I have to expect when I increase the operation timeout from 1 minute (the default) to 5, 10, or even 20 minutes?

Thanks for the advice and best,
Lukas

**Comment by Tembrel** [comment]
Is it sufficient for get() to be called on *some *future immediately to
prevent this from happening? I've been doing submitToAllMembers with
multi-minute tasks for a while, and I don't think I've seen this happen.

Would switching to the callback-based versions make this behavior less or
more likely?

**Comment by lukasblu** [comment]
Hi @Tembrel ,

"some" future might be enough - or not :-) I wrote another test which plays with a few combinations of waiting (or not) before calling Future.get() and calling Future.get() on the first (or the second) Future. So 4 combinations in total.

The test only succeeds if you call Future.get() immediately on the first Future. The other 3 combinations all fail.

```java
package com.nm.test.hazelcast.task;

import com.hazelcast.config.Config;
import com.hazelcast.config.XmlConfigBuilder;
import com.hazelcast.core.Hazelcast;
import com.hazelcast.core.HazelcastInstance;
import com.hazelcast.core.IExecutorService;
import com.hazelcast.core.Member;
import com.nm.test.hazelcast.TestHazelcast;
import org.apache.log4j.BasicConfigurator;
import org.apache.log4j.Logger;
import java.io.Serializable;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.concurrent.Callable;
import java.util.concurrent.Future;
import junit.framework.TestCase;

/**
 * Test if a task that runs longer than the call timeout is executed correctly when get is called
 * on the future after at least 2 times the call timeout has elapsed.
 * <p>
 * See https://github.com/hazelcast/hazelcast/issues/4398
 */
public class TestTask3 extends TestCase {

	private static final Logger logger = Logger.getLogger(TestTask3.class);

	private static final long timeoutMs = 3000;

	@Override
	protected void setUp() throws Exception {

		// configure logging
		if (!TestHazelcast.loggingInitialized) {
			TestHazelcast.loggingInitialized = true;
			BasicConfigurator.configure();
		}
	}

	public void testTask3NoSleep() throws Exception {
		testTask3(false, false);
	}

	public void testTask3NoSleepReverse() throws Exception {
		testTask3(false, true);
	}

	public void testTask3Sleep() throws Exception {
		testTask3(true, false);
	}

	public void testTask3SleepReverse() throws Exception {
		testTask3(true, true);
	}

	private void testTask3(boolean sleep, boolean reverse) throws Exception {

		// create hazelcast config
		Config config = new XmlConfigBuilder().build();
		config.setProperty("hazelcast.logging.type", "log4j");

		// set call timeout to 3 seconds to make the problem appear quicker
		config.setProperty("hazelcast.operation.call.timeout.millis", String.valueOf(timeoutMs));

		// start two hazelcast instances
		HazelcastInstance hcInstance1 = Hazelcast.newHazelcastInstance(config);
		HazelcastInstance hcInstance2 = Hazelcast.newHazelcastInstance(config);
		logger.info("Hazelcast instances started.");

		// var to collect execution results
		List<String> results = new ArrayList<String>();

		// try-finally to stop hazelcast
		try {

			// submit the long-running task
			logger.info("Executing task...");
			IExecutorService executorService = hcInstance1.getExecutorService("default");
			Map<Member, Future<String>> futures = executorService.submitToAllMembers(new SleepingTask());
			logger.info("Task submitted.");

			// sleep before calling Future.get() (2 x timeout + 2s to be after timeout)
			if (sleep) {
				Thread.sleep(2 * timeoutMs + 2000);
			}

			// wait for all results
			logger.info("Waiting for futures...");
			List<Future<String>> futuresList = new ArrayList<Future<String>>(futures.values());
			if (reverse) {
				Collections.reverse(futuresList);
			}
			for (Future<String> future : futuresList) {
				String result = future.get();
				results.add(result);
			}
			logger.info("Results received.");

		} catch (Exception e) {

			// log and rethrow exception
			logger.error("Exception: " + e.getMessage(), e);
			throw e;

		} finally {

			// shutdown
			hcInstance1.getLifecycleService().shutdown();
			hcInstance2.getLifecycleService().shutdown();
		}

		// ensure success
		for (String result : results) {
			if (!result.equals("Success")) {
				fail("Non-success result: " + result);
			}
		}

		logger.info("Test done.");
	}

	private static class SleepingTask implements Callable<String>, Serializable {

		private static final long serialVersionUID = 1L;

		@Override
		public String call() throws Exception {
			Thread.sleep(15000);
			return "Success";
		}

	}

}
```

To your second question: What do you mean by "callback-based versions"? Do you mean calling Future.isDone() and waiting until they are done before calling Future.get()? I have not tested that yet, but should be easy to do...

Best,
Lukas

**Comment by Tembrel** [comment]
On Wed, Feb 18, 2015 at 10:07 AM, Lukas Blunschi <notifications@github.com>
wrote:

> The test only succeeds if you call Future.get() immediately on the first
> Future.
>
OK, that explains why it seems to work for me.


> To your second question: What do you mean by "callback-based versions"? Do
> you mean calling Future.isDone() and waiting until they are done before
> calling Future.get()? I have not tested that yet, but should be easy to
> do...
>

No, I meant the MultiExecutionCallback form of submitToAllMembers.

http://docs.hazelcast.org/docs/3.4/javadoc/com/hazelcast/core/IExecutorService.html#submitToAllMembers(java.util.concurrent.Callable,
com.hazelcast.core.MultiExecutionCallback)


--tim

**Comment by Tembrel** [comment]
Sorry link broke over line. I meant this link
<http://docs.hazelcast.org/docs/3.4/javadoc/com/hazelcast/core/IExecutorService.html#submitToAllMembers(java.util.concurrent.Callable,
com.hazelcast.core.MultiExecutionCallback)>.

**Comment by lukasblu** [comment]
@Tembrel sure, no problem, I found the method.

I wrote another test which uses such a callback - and it again fails with an operation timeout :-(

In this case I see no workaround and therefore the method with a callback can simply not be used for long running tasks...

Best,
Lukas

```java
package com.nm.test.hazelcast.task;

import com.hazelcast.config.Config;
import com.hazelcast.config.XmlConfigBuilder;
import com.hazelcast.core.*;
import com.nm.test.hazelcast.TestHazelcast;
import org.apache.log4j.BasicConfigurator;
import org.apache.log4j.Logger;
import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.Callable;
import java.util.concurrent.CountDownLatch;
import junit.framework.TestCase;

/**
 * Test if a task that runs longer than the call timeout is executed correctly when get is called
 * on the future after at least 2 times the call timeout has elapsed.
 * <p>
 * See https://github.com/hazelcast/hazelcast/issues/4398
 */
public class TestTask4 extends TestCase {

	private static final Logger logger = Logger.getLogger(TestTask4.class);

	private static final long timeoutMs = 3000;

	@Override
	protected void setUp() throws Exception {

		// configure logging
		if (!TestHazelcast.loggingInitialized) {
			TestHazelcast.loggingInitialized = true;
			BasicConfigurator.configure();
		}
	}

	public void testTask4() throws Exception {

		// create hazelcast config
		Config config = new XmlConfigBuilder().build();
		config.setProperty("hazelcast.logging.type", "log4j");

		// set call timeout to 3 seconds to make the problem appear quicker
		config.setProperty("hazelcast.operation.call.timeout.millis", String.valueOf(timeoutMs));

		// start two hazelcast instances
		HazelcastInstance hcInstance1 = Hazelcast.newHazelcastInstance(config);
		HazelcastInstance hcInstance2 = Hazelcast.newHazelcastInstance(config);
		logger.info("Hazelcast instances started.");

		// var to collect execution results
		List<String> results = new ArrayList<String>();

		// try-finally to stop hazelcast
		try {

			// create callback
			CountDownLatch latch = new CountDownLatch(1);
			Callback callback = new Callback(latch);

			// submit the long-running task
			logger.info("Executing task...");
			IExecutorService executorService = hcInstance1.getExecutorService("default");
			executorService.submitToAllMembers(new SleepingTask(), callback);
			logger.info("Task submitted.");

			// wait for all results
			logger.info("Waiting for results...");
			callback.await();
			for (Object value : callback.values.values()) {
				results.add(value.toString());
			}
			logger.info("Results received.");

		} catch (Exception e) {

			// log and rethrow exception
			logger.error("Exception: " + e.getMessage(), e);
			throw e;

		} finally {

			// shutdown
			hcInstance1.getLifecycleService().shutdown();
			hcInstance2.getLifecycleService().shutdown();
		}

		// ensure success
		for (String result : results) {
			if (!result.equals("Success")) {
				fail("Non-success result: " + result);
			}
		}

		logger.info("Test done.");
	}

	private static class SleepingTask implements Callable<String>, Serializable {

		private static final long serialVersionUID = 1L;

		@Override
		public String call() throws Exception {
			Thread.sleep(15000);
			return "Success";
		}

	}

	private static class Callback implements MultiExecutionCallback {

		private final CountDownLatch latch;

		private Map<Member, Object> values;

		public Callback(CountDownLatch latch) {
			this.latch = latch;
		}

		@Override
		public void onResponse(Member member, Object value) {
			logger.info("Member " + member + " done: " + value);
		}

		@Override
		public void onComplete(Map<Member, Object> values) {
			this.values = values;
			logger.info("All members done: " + values.values());
			latch.countDown();
		}

		public void await() throws InterruptedException {
			latch.await();
		}

	}

}
```

**Comment by Tembrel** [comment]
If there's no way to change the operation timeout, then I'm glad that the
way I'm doing it -- calling get() immediately on the first of the futures
returned -- manages to work.

But maybe I should consider a design that uses ITopic to return the results
of long-running tasks.

**Comment by lukasblu** [comment]
@Tembrel 

Unfortunately, it looks like it is *not* always the first of the returned futures which is safe to call get() on. 

In my tests above it seams to be the same all the time, however in my application on one node it is the first, on the other the second. Basically, this renders the workaround useless, as you can not know which Future to call get() on.

So, I really have no idea how to safely execute long running tasks anymore. Increasing the operation call timeout might still be an option, but I would like to understand what negative consequences (if any) it could have to increase the timeout to something like 20 minutes.

Thanks and best,
Lukas

**Comment by lukasblu** [comment]
@mdogan

Your comment above is not fully correct. A better version would be:

"This happens only if more than 2 x timeout passes before calling future.get() on *all* returned futures."

This is quite a bit more serious, because my usual approach to wait for all answers is to call future.get() on one of the returned futures and after this answer has arrived continue to wait on the remaining futures:

```java

// execute task
IExecutorService executorService = hcInstance.getExecutorService(executorName);
Map<Member, Future<T>> futures = executorService.submitToAllMembers(task);

// wait for all results and return them in a collection
List<T> results = new ArrayList<T>();
for (Future<T> future : futures.values()) {
    T result = future.get();
    results.add(result);
}

```

If you get unlucky, this will throw an exception as shown below:

```java
2015-02-20 18:23:41,576 FATAL [hz.appway.response] com.hazelcast.spi.OperationService - [127.0.0.1]:5713 [blade] [3.4.1] While processing response...
com.hazelcast.core.HazelcastException: No invocation for response: NormalResponse{callId=1417, urgent=false, value=Slept 20s, backupCount=0}
	at com.hazelcast.spi.impl.BasicOperationService$ResponsePacketHandler.notifyRemoteCall(BasicOperationService.java:602)
	at com.hazelcast.spi.impl.BasicOperationService$ResponsePacketHandler.handle(BasicOperationService.java:586)
	at com.hazelcast.spi.impl.BasicOperationService$ResponsePacketHandler.access$1300(BasicOperationService.java:579)
	at com.hazelcast.spi.impl.BasicOperationService$BasicDispatcherImpl.dispatch(BasicOperationService.java:564)
	at com.hazelcast.spi.impl.BasicOperationScheduler$ResponseThread.process(BasicOperationScheduler.java:532)
	at com.hazelcast.spi.impl.BasicOperationScheduler$ResponseThread.doRun(BasicOperationScheduler.java:524)
	at com.hazelcast.spi.impl.BasicOperationScheduler$ResponseThread.run(BasicOperationScheduler.java:501)
```

This is quite bad and it would be great if you could look at this as soon as possible.

Thanks and best,
Lukas

**Comment by lukasblu** [comment]
Here is a short test that shows that executing the same long-running task on one node works while it fails on the other. It works on node 1 *or* on node 2, but never on both.

```java
package com.nm.test.hazelcast.task;

import com.hazelcast.config.Config;
import com.hazelcast.config.XmlConfigBuilder;
import com.hazelcast.core.Hazelcast;
import com.hazelcast.core.HazelcastInstance;
import com.hazelcast.core.IExecutorService;
import com.hazelcast.core.Member;
import com.nm.test.hazelcast.TestHazelcast;
import org.apache.log4j.BasicConfigurator;
import org.apache.log4j.Logger;
import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.Callable;
import java.util.concurrent.Future;
import junit.framework.TestCase;

/**
 * Test if a task that runs longer than the call timeout is executed correctly no matter on which
 * node Future.get() is called.
 * <p>
 * See https://github.com/hazelcast/hazelcast/issues/4398
 */
public class TestTask5 extends TestCase {

	private static final Logger logger = Logger.getLogger(TestTask5.class);

	private static final long timeoutMs = 3000;

	@Override
	protected void setUp() throws Exception {

		// configure logging
		if (!TestHazelcast.loggingInitialized) {
			TestHazelcast.loggingInitialized = true;
			BasicConfigurator.configure();
		}
	}

	public void testTask5() throws Exception {

		// create hazelcast config
		Config config = new XmlConfigBuilder().build();
		config.setProperty("hazelcast.logging.type", "log4j");

		// set call timeout to 3 seconds to make the problem appear quicker
		config.setProperty("hazelcast.operation.call.timeout.millis", String.valueOf(timeoutMs));

		// start two hazelcast instances
		HazelcastInstance hcInstance1 = Hazelcast.newHazelcastInstance(config);
		HazelcastInstance hcInstance2 = Hazelcast.newHazelcastInstance(config);
		logger.info("Hazelcast instances started.");

		// vars to collect execution results
		List<String> results1, results2;

		// try-finally to stop hazelcast
		try {

			// execute on node 1
			results1 = executeTask(hcInstance1, "node 1");

			// execute on node 2
			results2 = executeTask(hcInstance2, "node 2");

		} catch (Exception e) {

			// log and rethrow exception
			logger.error("Exception: " + e.getMessage(), e);
			throw e;

		} finally {

			// shutdown
			hcInstance1.getLifecycleService().shutdown();
			hcInstance2.getLifecycleService().shutdown();
		}

		// ensure success
		logger.info("Results from node 1: " + results1);
		logger.info("Results from node 2: " + results2);
		assertResults(results1);
		assertResults(results2);
	}

	private List<String> executeTask(HazelcastInstance hcInstance, String nodeName) throws Exception {
		List<String> results = new ArrayList<String>();

		// submit the long-running task
		logger.info("Executing long-running task...");
		IExecutorService executorService = hcInstance.getExecutorService("default");
		Map<Member, Future<String>> futures = executorService.submitToAllMembers(new SleepingTask(4 * timeoutMs + 1000));

		// wait for all results
		logger.info("Task submitted on " + nodeName + " - now waiting for results...");
		for (Future<String> future : futures.values()) {
			String result = future.get();
			results.add(result);
		}
		logger.info("Results received.");

		return results;
	}

	private void assertResults(List<String> results) {
		for (String result : results) {
			if (!result.equals("Success")) {
				fail("Non-success result: " + result);
			}
		}
	}

	private static class SleepingTask implements Callable<String>, Serializable {

		private static final long serialVersionUID = 1L;

		private final long sleepMs;

		public SleepingTask(long sleepMs) {
			this.sleepMs = sleepMs;
		}

		@Override
		public String call() throws Exception {
			Thread.sleep(sleepMs);
			logger.info("Slept for " + sleepMs / 1000 + "s.");
			return "Success";
		}

	}

}
```

**Comment by lukasblu** [comment]
Hi again (last update from my side for today:-)

I ran my tests above again with the latest version of the maintenance-3.x branch which contains the commit e4b93ea by @gurbuzali ("added a check before running execution-callback").

Unfortunately, the problem remains.

Cheers and have a nice weekend,
Lukas

**Comment by gurbuzali** [comment]
Hi @lukasblu, @Tembrel 
Thank you for all observations and tests, I'll take a look at them asap

**Comment by mdogan** [comment]
I've done a fix on my repo regarding this issue: https://github.com/mdogan/hazelcast/tree/issue-4398

See commit: https://github.com/mdogan/hazelcast/commit/a5db746d227b5ea0206e4f7f35025c0bd6c66cca

**Comment by lukasblu** [comment]
Hi,

thanks for fixing this problem.

However, there remains a problem when using callbacks to wait for completion of the task. I have opened a new issue #5564 to track this.

Thanks and best,
Lukas

**Comment by Sergey80** [comment]
on 3.3 all 2-5 tests are green. seems it is the most save version for now :)
